server/src/main/java/io/crate/planner/optimizer/rule/RewriteFilterOnCrossJoinToInnerJoin.java

```diff
@@ -68,24 +69,38 @@ public Pattern<Filter> pattern() {
     @Override
     public LogicalPlan apply(Filter equiJoinConditionFilter, Captures captures, Context context) {
         JoinPlan crossJoin = captures.get(joinCapture);
-        Symbol query = equiJoinConditionFilter.query();
-        Map<Set<RelationName>, Symbol> filterRelations = QuerySplitter.split(query);
 
-        List<RelationName> lhs = crossJoin.lhs().relationNames();
-        List<RelationName> rhs = crossJoin.rhs().relationNames();
-        // TODO this could be extended to also support nested-joins with more than 2 relations
-        if (lhs.size() == 1 && rhs.size() == 1) {
-            HashSet<RelationName> sources = HashSet.newHashSet(2);
-            sources.addAll(lhs);
-            sources.addAll(rhs);
-            if (filterRelations.containsKey(sources)) {
-                return new JoinPlan(
+        Map<Set<RelationName>, Symbol> filterQuery = QuerySplitter.split(equiJoinConditionFilter.query());
```

Comment:
This extends the cross-join to inner-join conversion to also support nested joins.